### DISCUSSION TITLE
Issues getting started

### DISCUSSION THREAD

**Discussion by ehuizar1028** [awslabs/soci-snapshotter#1267] [General]
I am using the getting started guide https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md

Which seems to be working as intended. I do see an improvement when i run an image on that single node.

But if i navigate to a different node where i didnt perform the `nerdctl pull/push` commands described in this section `push-an-image-to-your-registry` ,  but i install SOCI and pull with the soci snapshotter, i receive `unable to fetch SOCI index: unable to fetch descriptor` AND `mismatch content-length`

Soci and lazy loading is a new concept to me so I may be missing something fundamental, but does anyone have any ideas on what's going on?

**Comment by Kern--** [comment]
Sounds like a bug.

`mismatch content-length` comes from the `oras-go` library which we use for fetching SOCI indexes. It happens when the size reported in the `Content-Length` header of the response doesn't match what we expect. This shouldn't happen because we go to the registry to get the size before we request the data.

Can you provide additional information? What registry are you pulling from? Can you reproduce it with other images (this would be especially helpful if you can reproduce it with a public image so we can try on our side)? Can you provide additional SOCI logs?

**Reply by ehuizar1028** [reply]
Sure!

- Using Harbor as the image registry (i supposed i can try ECR as a test)
- I've tried 3 different images and all the same behavior
- Let me see what kind of logs i can attach here, unfortunately, my organization prohibits sending logs in a public forum, but given this is a test environment, let me see if I can get approval

**Reply by ehuizar1028** [reply]
@Kern-- 
Confirmed, this looks like it works as intended with ECR, but i still get the error with Harbor. 

My steps to reproduce is

Node1: Follow Guides
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#install-the-soci-snapshotter
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#push-an-image-to-your-registry (point to harbor endpoint instead of ECR)
- - Using `nvidia/cuda:12.4.1-cudnn-devel-ubuntu22.04`
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#run-container-with-the-soci-snapshotter
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#start-the-soci-snapshotter
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#run-container
- I can see the fuse mount, no errors in logs, fast startup. 

Node2: Follow Guides
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#install-the-soci-snapshotter
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#run-container-with-the-soci-snapshotter
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#start-the-soci-snapshotter
- https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#run-container
- It pulls/runs with the fallback, error shows `mismatch content-length`

**Reply by ehuizar1028** [reply]
I was able to reproduce the issue in an isolated environment outside my corporate environment on one node after following the clean slate procedure. 
https://github.com/awslabs/soci-snapshotter/blob/main/docs/debug.md#creating-a-clean-slate

Attached are the logs
[soci.log](https://github.com/user-attachments/files/15637428/soci.log)

I tried this in Harbor's demo environment since it's public, so you should be able to try this to reproduce. The demo environment gets wiped once a week, so if this can't be looked at before the wipe, then i can reupload or find an alternative. 

demo.goharbor.io/soci-ehuizar/soci-ehuizar/soci_cuda:latest

**Reply by Kern--** [reply]
I'm able to reproduce this using your image.

The descriptor that we resolve for the SOCI index looks weird. The size looks like the size of the image instead of the size of the index manifest (output from printing the object while debugging with delve):
```
github.com/opencontainers/image-spec/specs-go/v1.Descriptor {
	MediaType: "application/vnd.oci.image.manifest.v1+json",
	Digest: "sha256:c3afca17969ca3ea0f0bee7b77a985e62c877c080799f7fa65fc4d6a8...+7 more",
	Size: 67158202,
	URLs: []string len: 0, cap: 0, nil,
	Annotations: map[string]string [
		"com.amazon.soci.build-tool-identifier": "AWS SOCI CLI v0.1",
	],
	Data: []uint8 len: 0, cap: 0, nil,
	Platform: *github.com/opencontainers/image-spec/specs-go/v1.Platform nil,
	ArtifactType: "application/vnd.amazon.soci.index.v1+json",}
```

I also tried inspecting the response from fetching the content to get the real length:
```
	Header: net/http.Header [
		"Content-Type": [
			"application/vnd.oci.image.manifest.v1+json",
		],
		"Docker-Content-Digest": [
			"sha256:c3afca17969ca3ea0f0bee7b77a985e62c877c080799f7fa65fc4d6a8...+7 more",
		],
		"Content-Length": ["2488"],
	],
```


And overwriting the size in the descriptor via delve to the real length (2488) fixes the issue. Seems like something's wrong in Harbor.

**Reply by Kern--** [reply]
Oh actually, this is https://github.com/goharbor/harbor/issues/20187. Looks scheduled to release in v2.11

**Reply by Kern--** [reply]
Which was released 11 hours ago https://github.com/goharbor/harbor/releases/tag/v2.11.0 🎉

**Reply by ehuizar1028** [reply]
thank you! great timing then i guess! I have a dev instance i can update to 2.11 and will test and report back here. 

Thanks Again!

**Reply by Kern--** [reply]
Awesome. Let us know if that doesn't fix it and we can keep looking.

**Reply by ehuizar1028** [reply]
confirmed, it works with v2.11!

Thanks Again!
